File: sdk/js/packages/core/src/usecases/flush-buffer.ts
/**
 * Drains the buffer and delivers the batch to the gateway via Transport.
 * The only use case that performs I/O — called by the flusher timer and client.flush().
 */
import type { EventBuffer } from "../buffer.js";
import type { Transport } from "../ports/transport.js";

export class FlushBufferUseCase {
  constructor(
    private readonly buffer: EventBuffer,
    private readonly transport: Transport,
  ) {}

  /** Drain and ship. Returns the number of events sent (0 if buffer was empty). */
  async execute(): Promise<number> {
    const events = this.buffer.drain();
    if (events.length === 0) return 0;
    await this.transport.send(events);
    return events.length;
  }
}

File: sdk/python/src/watcher_sdk/usecases/flush_buffer.py
# Use case: drain the buffer and send the batch to the gateway via Transport.
# Runs on the background flusher thread — I/O allowed here.
from watcher_sdk.buffer import EventBuffer
from watcher_sdk.ports.transport import Transport


class FlushBufferUseCase:
    def __init__(self, buffer: EventBuffer, transport: Transport) -> None:
        self._buffer = buffer
        self._transport = transport

    def execute(self) -> int:
        """Drain the buffer and ship events. Returns the number of events sent."""
        events = self._buffer.drain()
        if not events:
            return 0
        self._transport.send(events)
        return len(events)
